File: Sample/tools/uart_loopback.py
#!/usr/bin/python

import serial
import sys, getopt
import time

def printUsage(prg) :
  print("\nUsage: " + prg + " [Options]\n")
  print("Options: -p <port>  (uart port)")
  print("         -b <baud>  (uart baud rate)")
  print("         -?         (show this help)")
  print("\nExample:")
  print("uart_loopback -p /dev/ttyACM0\n")
  sys.exit(1)

def main(argv) :
  baud = 115200
  tty = "/dev/ttyACM0"

  try:
    opts, args = getopt.getopt(argv, "p:b:?")
  except getopt.GetoptError as err:
    print(repr(err))
    sys.exit(1)

  for opt, arg in opts:
    if opt == '-?':
      printUsage(sys.argv[0])
    elif opt == "-b":
      baud = arg
    elif opt == "-p":
      tty = arg

  print("Uart port used: " + tty + " baud: " + "{}".format(baud))

  ser = serial.Serial(
                    port=tty, \
                    baudrate=baud, \
                    parity=serial.PARITY_NONE, \
                    stopbits=serial.STOPBITS_ONE, \
                    bytesize=serial.EIGHTBITS, \
                    timeout=0, \
                    write_timeout=0)

  while True :
    if ser.inWaiting() > 0 :
      byteReads = ser.read(ser.inWaiting())
      ser.write(byteReads)
      for b in byteReads :
        print("{" + "{:02X}".format(b) + "}", end ='', flush=True)

if __name__ == "__main__":
  main(sys.argv[1:])

File: Sample/tools/uart_loopback.c
/*
 * uart_loopback.c - userspace uart loopback test
 *
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 *
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <getopt.h>
#include <termios.h>
#include <sys/ioctl.h>

void print_usage(char *prg) {
    fprintf(stderr, "\nUsage: %s [options] tty\n\n", prg);
    fprintf(stderr, "Options: -b <baud>  (uart baud rate)\n");
    fprintf(stderr, "         -?         (show this help)\n");
    fprintf(stderr, "\nExample:\n");
    fprintf(stderr, "uart_loopback /dev/ttyS1\n");
    exit(1);
}

int main(int argc, char **argv) {
    fd_set readfs;    /* file descriptor set */
    struct timeval Timeout;
    int fd, ret;
    char *tty;
    uint8_t j;
    char *btr = NULL;
    int opt,count=127;
    char i;

    while ((opt = getopt(argc, argv, "b:?")) != -1) {
        switch (opt) {
        case 'b':
            btr = optarg;
            printf("Baud=%s\n",btr);
            break;

        case '?':
        default:
            print_usage(argv[0]);
            break;
        }
    }

    if (argc - optind != 1)
        print_usage(argv[0]);

    tty = argv[optind];

    printf("uart port used: %s\n", tty);

    if ((fd = open (tty, O_RDWR | O_NOCTTY)) < 0) {
        perror(tty);
        exit(1);
    }

    tcflush(fd, TCIFLUSH);

    while (1) {
        FD_SET(fd, &readfs);  /* set testing source */

        /* set timeout value within input loop */
        Timeout.tv_usec = 1;  /* milliseconds */
        Timeout.tv_sec  = 0;  /* seconds */
        ret = select(fd + 1, &readfs, NULL, NULL, &Timeout);
        if (ret == 0) {
            // printf("read timeout error!\n");
            // exit(1);
        } else
            ret = read(fd, &j, 1);

        if (ret != 1) {
            // printf("read error!\n");
            // exit(1);
        } else {
            ret = write(fd, &j, 1);
            if (ret != 1) {
                printf("write error!\n");
                exit(1);
            }
            printf("{%.2X}", j);
            fflush(stdout);
        }
    }

    printf("\n");

    close(fd);

    return 0;
}